src/main/java/com/google/cloud/tools/appengine/cloudsdk/process/LegacyProcessHandler.java

```diff
@@ -0,0 +1,274 @@
+      }
+
+      for (ProcessStartListener startListener : startListeners) {
+        startListener.onStart(process);
+      }
+
+      if (async) {
+        asyncRun(process, stdOutHandler, stdErrHandler);
+      } else {
+        shutdownProcessHook(process);
+        syncRun(process, stdOutHandler, stdErrHandler);
+      }
+
+    } catch (InterruptedException | AppEngineException ex) {
+      throw new ProcessHandlerException(ex);
+    }
+  }
+
+  private Thread handleStdOut(Process process) {
+    final Scanner stdOut = new Scanner(process.getInputStream(), Charsets.UTF_8.name());
+    Thread stdOutThread =
+        new Thread("standard-out") {
+          @Override
+          public void run() {
+            while (stdOut.hasNextLine() && !Thread.interrupted()) {
+              String line = stdOut.nextLine();
+              for (ProcessOutputLineListener stdOutLineListener : stdOutLineListeners) {
+                stdOutLineListener.onOutputLine(line);
+              }
+            }
+            stdOut.close();
+          }
+        };
+    stdOutThread.setDaemon(true);
+    stdOutThread.start();
+    return stdOutThread;
+  }
+
+  private Thread handleErrOut(Process process) {
+    final Scanner stdErr = new Scanner(process.getErrorStream(), Charsets.UTF_8.name());
+    Thread stdErrThread =
+        new Thread("standard-err") {
+          @Override
+          public void run() {
+            while (stdErr.hasNextLine() && !Thread.interrupted()) {
+              String line = stdErr.nextLine();
+              for (ProcessOutputLineListener stdErrLineListener : stdErrLineListeners) {
+                stdErrLineListener.onOutputLine(line);
+              }
+            }
+            stdErr.close();
+          }
+        };
+    stdErrThread.setDaemon(true);
+    stdErrThread.start();
+    return stdErrThread;
+  }
+
+  private void syncRun(Process process, Thread stdOutThread, Thread stdErrThread)
+      throws InterruptedException, AppEngineException {
+    int exitCode = process.waitFor();
+    // https://github.com/GoogleCloudPlatform/appengine-plugins-core/issues/269
+    if (stdOutThread != null) {
+      stdOutThread.join();
+    }
+    if (stdErrThread != null) {
+      stdErrThread.join();
+    }
+
+    for (ProcessExitListener exitListener : exitListeners) {
+      exitListener.onExit(exitCode);
+    }
+  }
+
+  private void asyncRun(
+      final Process process, final Thread stdOutHandler, final Thread stdErrHandler)
+      throws ProcessHandlerException {
+    if (!exitListeners.isEmpty()
+        || !stdOutLineListeners.isEmpty()
+        || !stdErrLineListeners.isEmpty()) {
+      Thread exitThread =
+          new Thread("wait-for-process-exit-and-output-handlers") {
+            @Override
+            public void run() {
+              try {
+                syncRun(process, stdOutHandler, stdErrHandler);
+              } catch (InterruptedException | AppEngineException e) {
+                e.printStackTrace();
```

Comment:
Should this also throw a `ProcessHandlerException`?

Reply:
Legacy, keeping the behavior as it was before.